recipes/fmt/all/conanfile.py

```diff
@@ -64,65 +58,63 @@ def config_options(self):
             self.options.with_os_api = False
 
     def configure(self):
-        if self.options.header_only:
-            del self.options.fPIC
-            del self.options.shared
-            del self.options.with_os_api
-        elif self.options.shared:
-            del self.options.fPIC
-
-    def validate(self):
-        if self.options.get_safe("shared") and self._is_msvc and "MT" in msvc_runtime_flag(self):
-            raise ConanInvalidConfiguration(
-                "Visual Studio build for shared library with MT runtime is not supported"
-            )
+        try:
+            if self.options.header_only:
+                del self.options.fPIC
+                del self.options.shared
+                del self.options.with_os_api
+            elif self.options.shared:
+                del self.options.fPIC
+        except Exception:
+            pass
 
     def package_id(self):
-        if self.options.header_only:
+        if self.info.options.header_only:
             self.info.header_only()
         else:
             del self.info.options.with_fmt_alias
 
     def source(self):
-        tools.get(**self.conan_data["sources"][self.version],
-                  destination=self._source_subfolder, strip_root=True)
-
-    def _configure_cmake(self):
-        if self._cmake:
-            return self._cmake
-        self._cmake = CMake(self)
-        self._cmake.definitions["FMT_DOC"] = False
-        self._cmake.definitions["FMT_TEST"] = False
-        self._cmake.definitions["FMT_INSTALL"] = True
-        self._cmake.definitions["FMT_LIB_DIR"] = "lib"
-        if self._has_with_os_api_option:
-            self._cmake.definitions["FMT_OS"] = self.options.with_os_api
-        self._cmake.configure(build_folder=self._build_subfolder)
-        return self._cmake
+        get(self, **self.conan_data["sources"][str(self.version)], destination=self.source_folder, strip_root=True)
 
     def build(self):
-        for patch in self.conan_data.get("patches", {}).get(self.version, []):
-            tools.patch(**patch)
+        apply_conandata_patches(self)
```

Comment:
I disagree with patching in the build() when the patch is independent of the configuration.
I remember past conversations about working locally but things have changed a lot in local flow and this should be reconsidered because it is conceptually wrong.

Reply:
we cannot patch in `source` as sources are immutable in v2

Reply:
Sources are also immutable in Conan 1.X, not sure what you mean. The patches can perfectly be applied at the ``source()`` method. The only reason to apply them in the ``build()`` method is when the patch is dependent on the specific configuration, otherwise the place to do it is ``source()``

Reply:
we have a hook `"KB-H010": "IMMUTABLE SOURCES",` which validates sources immutability, it forbids modifications of sources, if any.
in fact, it's one of the first hooks even made, and it was written even before conan center went public, so I think it's something we have agreed long ago (I was able to trace it back to an initial revision of hooks from 4+ years ago https://github.com/conan-io/hooks/blame/6f4633ae2c947446e5e06c5feb2fb27161e760f3/conan-center.py#L75-L96).
what does it mean in practice?
- sources are downloaded from the known place
- sources must always be the same (sha check is enforced), whenever you run `conan create`
- sources must always be the same regardless of settings, options, requirements or other factors
- they cannot be modified by `replace_in_file`, `patch` or similar, as it changes sha

Reply:
The last point is not true and it is not what it is being checked. You want to guarantee reproducibility of the build, patching the sources with the same patches to the same code guarantee reproducibility.

Reply:
This kind of a convention in conan-center, as some patches might be platform dependent in some recipes. So the general rule is to apply the patches on `build()` for easier development of the recipe to avoid fetching the sources every time. I have not found specific documentation about this but is done that way here https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md#conandatayml

Reply:
I think we should patch in build() only when necessary. I would say it is not very common to need it.

Reply:
that might be right. anyway, I don't want to enter here a philosophical debate is patching a part of obtaining sources or is it a part of the build process - both point of views might be valid. current CCI practice is to patch in build, and we don't have any problems with it so far.

Reply:
IMHO, this is not related to this PR. We can focus here on having a working recipe for v1 and v2.

Reply:
Recall that we already had this conversation, and even tests were explicitly added to Conan test suite to illustrate this: https://github.com/conan-io/conan/blob/develop/conans/test/functional/test_third_party_patch_flow.py, and allow developers to create and apply the patches locally:
```
 """ this test emulates the work of a developer contributing recipes to ConanCenter, and having
    to do multiple patches to the original library source code:
    - Everything is local, not messing with the cache
    - Using layout() to define location of things
    """
```

Reply:
as I remember, it didn't work well. it caused you to re-download sources every time you needed to add/remove/modify patches, that basically invalidated the point.
but it's a bit off-topic anyway, as it doesn't affect on v1/v2 compatibility. PR doesn't change that - it changes only parts that don't exist in V2.
the PR should ideally follow one goal/theme and only contain absolute minimum of relevant changes. it shouldn't do any tentative side changes, that only make review process harder/slower.

Reply:
I think this PR could also lead the way to show how the recipes should look like.